### PR TITLE
Improve default logger for realm/react

### PR DESCRIPTION
**Pull request by takameyer**
## What, How & Why?
The default logger could be a bit smarter.  This will print a contextual log message with the level and use either `console.error`, `console.warn` and `console.log` based on the relevant level.  This also sets a default log level to WARN.

## ☑️ ToDos

* [ ] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [ ] 📝 Update `COMPATIBILITY.md`
* [ ] 🚦 Tests
* [ ] 🔀 Executed flexible sync tests locally if modifying flexible sync
* [ ] 📦 Updated internal package version in consuming `package.json`s (if updating internal packages)
* [ ] 📱 Check the React Native/other sample apps work if necessary
* [ ] 📝 Public documentation PR created or is not necessary
* [ ] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [ ] typescript definitions file is updated
* [ ] jsdoc files updated

### CODE REVIEW

**Comment by kraenhansen** [comment]
Why not add this to the `realm` package directly? I think a good default logger would be valuable more generally.

**Comment by takameyer** [comment]
> Why not add this to the realm package directly?

I've wanted to do this for a while, so I have no argument against not doing it directly.

**Comment by takameyer** [comment]
> @realm/react won't override the default

The unified logger will be at a different entry point and this would turn into a no-op.  So this will just be backwards compatible. I want to still ensure v11 users get the advantages of the `@realm/react` updates.

**Comment by kraenhansen** [comment]
As I read the last comment on the unified logger issue, the `Realm.App.Sync.setLogger` will change to call `Realm.setLogger` and with this PR merged `@realm/react` will override whatever default the `realm` package injects on import, right? What am I missing? To me this change will interfere with the future default behavior of `realm@v12`.

**Comment by takameyer** [comment]
@kraenhansen How would you recommend we support both versions of logging?  With this change, users wouldn't even have to update `realm` to get default logger.  After unified logging is realised, we can make this support both. 

> Realm.App.Sync.setLogger will change to call Realm.setLogger

This will be a breaking change then right?  We should probably make this a v13 change if `Realm.App.Sync.setLogger` doesn't work anymore.

File: packages/realm-react/src/AppProvider.tsx [hunk 1]
```diff
@@ -36,8 +36,49 @@ type AppProviderProps = Realm.AppConfiguration & {
   logger?: (level: Realm.App.Sync.NumericLogLevel, message: string) => void;
 };
 
+// This is the default log level that would be useful for React Native.
+const DEFAULT_LOG_LEVEL: LogLevel = "warn";
+
+// Since the logger provides the numeric log level, we need to convert it to a
+// string for the log message.
+const getLogLevelString = (level: Realm.App.Sync.NumericLogLevel) => {
```
Comment (elle-j) [inline packages/realm-react/src/AppProvider.tsx:44]: `NumericLogLevel.All` and `NumericLogLevel.Off` as well 👍

**Review by elle-j — APPROVED** [review]
I like the split off the various console methods!

**Comment by takameyer** [comment]
I may be closing this, as the effort @papafe is currently doing will probably nullify this, as well as the `logLevel` and `logger` properties of the `AppProvider`.  Part of me does want to release this, just so people that haven't updated to `v12` can reap the benefits, although I'm unsure if this comes into conflict with the recent core versions.

**Pull request closed without merging**